### PR TITLE
Local: mimic the path semantics of cloud backends

### LINKED ISSUE #170: Local Backend: Creating Files from Containers Ignores Container Path
If I create a local container, then do

```
container.Item("somefile.txt")
```

it looks for a file relative to my current working directory called "somefile.txt" instead of looking relative to the container folder. 

This line: https://github.com/graymeta/stow/blob/master/local/container.go#L124

sets the item path

```
path := id
```

but based on the behavior of other containers (at least S3) I would expect this to be 

```
path := filepath.Join(c.path, id)
```

unless I'm misunderstanding something.

### PR DESCRIPTION
**Pull request by urisimchoni**
This fixes #170

Store the Item's relative path to its container, not
its full path. When constructing container-based items,
treat relative paths as relative to the container, but keep
supporting absolute paths as well.

This is analogous to cloud backends, and so **container.Item("name")**
means the same thing in all backends - it means an item called "name"
that is stored inside the container.

On the other hand, we still want to support
container.Item("/full/path/to/name"), because the full path is the ID
of the Item. So this patch interprets an absolute path as the full
path of the item, and converts that to a relative path for internal
storage.

### CODE REVIEW

**Comment by urisimchoni** [comment]
PR updated with a commit to do the same with containers, relative to their location (i.e. for the local backend, location.Container("foo") means /path/to/location/foo)

**Comment by kumare3** [comment]
@urisimchoni are you thinking of pushing this? Is something blocking you? I encountered this `bug` too

**Comment by urisimchoni** [comment]
@kumare3 I have no push rights here. Things have been pretty quiet in this project - we need to reach the maintainers and ask what's up.

**Comment by johnmotz** [comment]
@urisimchoni Team is reviewing now along with your other PRs.

File: local/container.go [hunk 1]
```diff
@@ -31,11 +31,16 @@ func (c *container) URL() *url.URL {
 	}
 }
 
-func (c *container) CreateItem(name string) (stow.Item, io.WriteCloser, error) {
-	path := filepath.Join(c.path, name)
+func (c *container) pathAndItem(name string) (string, *item) {
```
Comment (jsteenb2) [inline local/container.go]: not sure crazy about this method, why not just do it inline?

File: local/container.go [hunk 2]
```diff
@@ -105,33 +108,41 @@ func (c *container) Items(prefix, cursor string, count int) ([]stow.Item, string
...
-	info, err := os.Stat(path)
+	var fullPath string
+	if filepath.IsAbs(id) {
+		//Absolute path - id has native OS semantics
+		fullPath = id
+	} else {
+		//relative path - id has '/' semantics
+		fullPath = filepath.Join(c.path, filepath.FromSlash(id))
+	}
+	info, err := os.Stat(fullPath)
```
Comment (jsteenb2) [inline local/container.go]: the err could be something other than `os.IsNotExist`, and we will want to check on that as well

File: local/item.go [hunk 3]
```diff
@@ -27,19 +27,24 @@ const (
 )
 
 type item struct {
-	path     string
-	infoOnce sync.Once // protects info
-	info     os.FileInfo
-	infoErr  error
-	metadata map[string]interface{}
+	relPath   string // OS-native relative path
+	container *container
```
Comment (jsteenb2) [inline local/item.go]: why put the entire container on the item instead of just its path? A lot of coupling here that seems unwarranted

File: local/location.go [hunk 4]
```diff
@@ -19,9 +20,27 @@ func (l *location) Close() error {
 }
 
 func (l *location) ItemByURL(u *url.URL) (stow.Item, error) {
-	i := &item{}
-	i.path = u.Path
-	return i, nil
+	//The URL is an absolute path. It can be broken in many ways into
```
Comment (jsteenb2) [inline local/location.go]: we'll need an additional test case that outlines why this change is needed

File: local/location.go [hunk 5]
```diff
@@ -104,7 +123,16 @@ func (l *location) Container(id string) (stow.Container, error) {
 	if !ok {
 		return nil, errors.New("missing " + ConfigKeyPath + " configuration")
 	}
-	containers, err := l.filesToContainers(path, id)
+	var fullPath string
+	if filepath.IsAbs(id) {
+		//Absolute path - id has native OS semantics
```
Comment (jsteenb2) [inline local/location.go]: comments aren't adding any value in the if/else block here, obvious from the code itself

Comment (jsteenb2) [reply local/location.go]: axe em

File: local/item.go [hunk 6]
```diff
@@ -27,19 +27,24 @@ const (
 )
 
 type item struct {
-	path     string
-	infoOnce sync.Once // protects info
-	info     os.FileInfo
-	infoErr  error
-	metadata map[string]interface{}
+	relPath   string // OS-native relative path
```
Comment (jsteenb2) [inline local/item.go]: axe the comment

File: local/item.go [hunk 7]
```diff
@@ -27,19 +27,24 @@ const (
...
+	container *container
+	infoOnce  sync.Once // protects info
+	info      os.FileInfo
+	infoErr   error
+	metadata  map[string]interface{}
+}
+
+func (i *item) fullPath() string {
```
Comment (jsteenb2) [inline local/item.go]: move unexported methods after the exported methods for the `item` type

File: local/item.go [hunk 8]
```diff
@@ -92,7 +97,7 @@ func (i *item) ensureInfo() error {
 }
 
 func (i *item) setMetadata(info os.FileInfo) error {
-	fileMetadata := getFileMetadata(i.path, info) // retrieve file metadata
+	fileMetadata := getFileMetadata(i.fullPath(), info) // retrieve file metadata
```
Comment (jsteenb2) [inline local/item.go]: do us a solid and remove this comment too, code says it all

File: local/container.go [hunk 9]
```diff
@@ -105,33 +108,41 @@ func (c *container) Items(prefix, cursor string, count int) ([]stow.Item, string
...
 	}
 	return items, cursor, nil
 }
 
 func (c *container) Item(id string) (stow.Item, error) {
-	path := id
-	info, err := os.Stat(path)
+	var fullPath string
+	if filepath.IsAbs(id) {
+		//Absolute path - id has native OS semantics
```
Comment (jsteenb2) [inline local/container.go]: can lose the comments for the if/else blocks here, code says it all

File: local/container.go [hunk 10]
```diff
@@ -105,33 +108,41 @@ func (c *container) Items(prefix, cursor string, count int) ([]stow.Item, string
...
-		if err != nil {
-			return nil, "", err
-		}
 		if !strings.HasPrefix(f.Name(), prefix) {
 			continue
 		}
 		item := &item{
-			path: path,
+			relPath:   f.Name(),
+			container: c,
```
Comment (jsteenb2) [inline local/container.go]: not digging the full coupling of the container to the item

File: local/container.go (hunk 1 above)
Comment (jsteenb2) [reply local/container.go]: Also, the unexported methods should go after all the exported methods for the `container` type

**Comment by urisimchoni** [comment]
New version. This is pretty much a rewrite of the previous PR:
- no link from item to container, instead keep the length of the path that belongs to the component
- the added tests are generic tests instead of local-specific. This demonstrates the behavior in other backends, namely that c.Item(name) is relative to the container etc. - the new tests pass with cloud backends, but only pass on the local backend with changes applied
- add a local-specific test to document the URL-to-item tranlsation, as requested in the PR.

**Comment by ajbouh** [comment]
Any update on this?
